[synthetics] Add projects workflow docs (#2353)
elastic/observability-docs: docs/en/observability/synthetics-projects-workflow.asciidoc

[[synthetics-project-monitors]]
// lint ignore params
= Working with project monitors

++++
<titleabbrev>Working with project monitors</titleabbrev>
++++

beta[]

Projects are the most powerful and sophisticated way to configure synthetic monitors in the Elastic Stack. Projects let you define your infrastructure as code, more commonly known as IaaC or git-ops. With project monitors you organize your YAML configuration and Javascript/Typescript defined monitors on the filesystem, use Git for version control, and deploy via a CLI tool, usually executed on a CI/CD platform.

> == Getting started

Reviewer: Did you intend for each section to be on its own page (see [the preview](https://observability-docs_2353.docs-preview.app.elstc.co/guide/en/observability/master/synthetics-project-monitors.html))? If not, you can add a `[discrete]` tag above the h2 headings (`==`).

```suggestion
[discrete]
== Getting started
```

Author: Not at all! This is much better

Working with projects requires working with the Elastic Synthetics CLI tool, which
can be invoked via the `npx @elastic/synthetics` command. Before getting started
you'll need to install Node.js; the `npx` command is provided by node, downloading
the `@elastic/synthetics` NPM module if not yet available on your system.

Run `npx @elastic/synthetics -h` to display the help menu and confirm your

> system is setup correctly

Author: I'm not certain where you're seeing a list being used here.

> Working with projects requires working with the Elastic Synthetics CLI tool, which
> can be invoked via the `npx @elastic/synthetics` command. Before getting started
> you'll need to install Node.js; the `npx` command is provided by node, downloading
> the `@elastic/synthetics` NPM module if not yet available on your system.
> 
> Run `npx @elastic/synthetics -h` to display the help menu and confirm your
> system is setup correctly

Reviewer: Based on what's in the [elastic/synthetics README](https://github.com/elastic/synthetics#usage), I would say:

```suggestion
Working with projects requires working with the Elastic Synthetics CLI tool, which
can be invoked via the `npx @elastic/synthetics` command. Before getting started
you'll need to:

. Install https://nodejs.dev/en/[Node.js]
. Install the package:
+
[source,sh]
----
npm install -g @elastic/synthetics
----
. To display the help menu and confirm your system is setup correctly, run:
+
[source,sh]
----
npx @elastic/synthetics -h
----
```

But based on this:

>the `npx` command is provided by node, downloading
>the `@elastic/synthetics` NPM module if not yet available on your system.

I'm not sure if the second step is necessary.

It is highly recommended to use an https://cloud.elastic.co/[Elastic Cloud] deployment with this tutorial.
If you wish to use a self-managed deployment, please read the <<synthetics-private-location>> docs first, and setup a private location with Fleet before proceeding.

We'll start by creating our first project, run the command below to create a new 
project named `projects-test` in the current directory.

`npx @elastic/synthetics init projects-test`

Then, follow the prompts on screen to setup the correct default variables for your project.
When complete, set the `SYNTHETICS_API_KEY` environment variable in your terminal, which is used
for authentication with your Elastic Stack. You will most likely want to set this permanently. 
This is done differently for https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.2#saving-changes-to-environment-variables[Powershell] and https://unix.stackexchange.com/a/117470[Bash].

> Let's pause for a moment and examine the key files and directories inside your project. 
> 
> * `journeys` contains the `.yaml`, `.ts` and `.js` files defining your monitors.
> * `synthetics.config.ts` contains settings for your project
> * `package.json` contains NPM settings for your project. Learn more in the https://docs.npmjs.com/about-packages-and-modules[NPM documentation].
> * `.github` contains sample Github actions workflow files for use with Github Actions.

Reviewer: Use second person.

Also, what do you think about being explicit about the sample monitor code included when you create a new project?  We _could_ also add a link to [Synthetic tests configuration](https://www.elastic.co/guide/en/observability/current/synthetics-configuration.html) to "learn more" (like you did for "NPM documentation"), _but_ I'm not sure you want to encourage the reader to click over there just yet.

```suggestion
Then, take a look at key files and directories inside your project:

* `journeys` is where you'll add `.yaml`, `.ts`, and `.js` files defining your monitors. When you create a new project, this directory will contain files defining sample monitors.
* `synthetics.config.ts` contains settings for your project. When you create a new project, it will contain some basic configuration options that you can customize later.
* `package.json` contains NPM settings for your project. Learn more in the https://docs.npmjs.com/about-packages-and-modules[Synthetic tests configuration].
* `.github` contains sample Github actions workflow files for use with Github Actions.
```

Author: I agree, I'd rather not sidetrack the user here, and they can easily get there by clicking around.

> While inside our project directory we can do two things with the `npx @elastic/synthetics` command:
> 
> * Run browser based monitors defined in `.ts` and `.js` files)
> * Push configurations to 
> 
> Let's try running the example journeys in the `journeys` folder by executing the following command.
> 
> `npx @elastic/synthetics journeys`
> 
> Will run the browser journeys locally.
> 
> Next, let's try pushing our journeys to our Elastic deployment by running the following command
> from inside your project.
> 
> `npx @elastic/synthetics push`

Reviewer: Use second person. 

Also, I think this is a bit repetitive. We could condense the list and the code blocks:

```suggestion
While inside the project directory you can do two things with the `npx @elastic/synthetics` command:

* Test browser-based monitors locally. To run all journeys defined in `.ts` and `.js` files:
+
[source,sh]
----
npx @elastic/synthetics journeys
----
* Push all monitor configurations to an Elastic deployment. Run the following command from inside your project:
+
[source,sh]
----
npx @elastic/synthetics push
----
```

Looks like:

<img width="860" alt="Screen Shot 2022-11-09 at 3 12 34 PM" src="https://user-images.githubusercontent.com/10479155/200942856-00b09c42-4951-4b0b-b8f8-a606c30f38d0.png">

Author: Much better

> Now, visit the Uptime UI in Kibana, you should see your newly pushed monitors running, and their
> config settings visible under the 'Monitor Management' section.
> 
> You've now setup a project, run monitors locally, and pushed them remotely to an Elastic Stack instance.

Reviewer: Break this into shorter, simpler sentences.

Also, I think this process is short enough that you don't need a summary at the end.

```suggestion
Then, go to the Uptime UI in Kibana. You should see your newly pushed monitors running. You can also go to the *Monitor Management* page to see the monitors' configuration settings.
```

== Recommended workflow

> Now that you have a working project, let's cover operationalizing it. You'll want to version control all
> files in git. If your project is not already in a version controlled directory add it now, and push it to
> your git host.

Reviewer: I'm not sure about the use of the word `operationalizing` here. It seems vague to me (unless it's an industry term I'm not aware of). I'm also not sure this suggestion is right, but maybe something like:

```suggestion
After your project is set up, there are some best practices you can implement to manage project monitors effectively.

First, it's recommended that you version control all files in git. If your project is not already in a version controlled directory add it now, and push it to your git host.
```

Author: I like this wording, yeah it's a weird use of a possibly fake word.

> While it can be convenient, especially when setting up a new project, to run the `push` command directly
> from your workstation, it's generally considered inadvisable to do so in a production environment. 
> Our recommended best practice is to:
> 
> 1. Develop and test changes locally
> 2. Create a pull request for all config changes
> 3. Have your CI service automatically very the PR by running `npx @elastic/synthetics .`
> 4. Have a human approve the pull request
> 5. Merge the pull request
> 6. Have your CI service automatically deploy the change by running `npx @elastic/synthetics push` after merge.
> 
> Given the considerable differences in CI systems and Git hosts the implementation of these steps is
> mostly left as an exercise to the reader. We do include a sample Github configuration in the `.github` 
> folder in your generated project.

Reviewer: I think there are opportunities to simplify the language here.

```suggestion
While it can be convenient to run the `push` command directly from your workstation,
especially when setting up a new project, it is not recommended for production environments.

Instead, we recommended that you:

1. Develop and test changes locally
2. Create a pull request for all config changes
3. Have your CI service automatically verify the PR by running `npx @elastic/synthetics .`
4. Have a human approve the pull request
5. Merge the pull request
6. Have your CI service automatically deploy the change by running `npx @elastic/synthetics push` after changes are merged

The exact implementation details will depend on the CI system and Git host you use. 
You can reference the sample Github configuration file that is included in the `.github`
directory when you create a new project.
```
